bokeh/io.py

```diff
@@ -390,19 +391,29 @@ def save(obj, filename=None, resources=None, title=None, state=None, validate=Tr
 
 def _detect_filename(ext):
     """ Detect filename from the name of the script being run. Returns
-    None if the script could not be found (e.g. interactive mode).
+    temporary file if the script could not be found or the location of the
+    script does not have write permission (e.g. interactive mode).
     """
     import inspect
-    from os.path import isfile, dirname, basename, splitext, join
+    from os.path import dirname, basename, splitext, join, curdir
 
     frame = inspect.currentframe()
     while frame.f_back and frame.f_globals.get('name') != '__main__':
         frame = frame.f_back
 
     filename = frame.f_globals.get('__file__')
-    if filename and isfile(filename):
-        name, _ = splitext(basename(filename))
-        return join(dirname(filename), name + "." + ext)
+
+    if filename is None:
+        return tempfile.NamedTemporaryFile().name
+
+    # dirname(filename) returns empty str if file called from within directory
```

Comment:
I can confirm that that this function returns `"foo.py"` if run e.g as `python foo.py` and `"somedir/foo.py"` if run as `python somedir/foo.py` in which in the former case the dirname ends up returning the empty string. I can also confirm the `os.access` does not interpret the empty string as the current dir:
```
In [9]: os.access(curdir, os.W_OK | os.X_OK)
Out[9]: True

In [10]: os.access("", os.W_OK | os.X_OK)
Out[10]: False
```
So I think testing against `curdir` makes sense.